### PR TITLE
Fix Issue 18367 - dmd should not segfault on -X with libraries, but no source files

### PR DESCRIPTION
**Pull request by wilzbach**
I wish all segfaults would be that easy.

### CODE REVIEW

File: src/dmd/mars.d [hunk 1]
```diff
@@ -880,6 +880,9 @@ private int tryMain(size_t argc, const(char)** argv)
             }
             else
             {
+                if (global.params.objfiles.dim == 0)
```
Comment (UplinkCoder) [inline src/dmd/mars.d]: @wilzbach indentation seems to suggest you missed braces.

**Review by UplinkCoder — CHANGES_REQUESTED** [review]

**Review by UplinkCoder — APPROVED** [review]
LGTM

**Comment by wilzbach** [comment]
Argh this is failing on Windows :/

**Comment by timotheecour** [comment]
https://github.com/dlang/dmd/pull/7838 will supersede this behavior right? (but still good to submit in meantime)

**Comment by wilzbach** [comment]
I thought you don't want `-X` to emit JSON, but just `-Xf=-`?

**Comment by timotheecour** [comment]
ya, sorry ignore my comment

**Comment by marler8997** [comment]
PR https://github.com/dlang/dmd/pull/7838 addresses this.  I defaulted to writing to `info.json`, however, your approach of asserting an error works as well.  Let me know what you want to do.

**Comment by wilzbach** [comment]
I think producing an error is saner then writing to a default info.json file. This would just need extra documentation and the case isn't very popular anyways.

**Comment by marler8997** [comment]
Ok I changed my PR to issue an error in this case:
```
Error: cannot determine JSON filename, specify using -Xf=<file> or provide at least one source file
```

File: test/compilable/test18367.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (marler8997) [inline test/compilable/test18367.sh]: bash wizadry

Comment (marler8997) [reply test/compilable/test18367.sh]: I copied this test to my PR, and added another case to it

**Comment by timotheecour** [comment]
> I defaulted to writing to info.json, however, your approach of asserting an error works as well. Let me know what you want to do.

indeed, IMO as a general rule it's always better to be explicit (let user choose file name or give error); it reduces bugs (eg what if a user already had an info.json and didn't intend it to be deleted)

**Comment by wilzbach** [comment]
Closing this in favor of #7838 then. Thanks @marler8997 !

**Pull request closed without merging**